fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintenanceTest.java

```diff
@@ -781,6 +784,85 @@ public void ensureValid() throws IOException {
+        // Synchronization blocks in FDBDirectoryWrapper can cause a deadlock
+        // see https://github.com/FoundationDB/fdb-record-layer/issues/2989
+        // So set the pool large enough to overcome that
+        this.dbExtension.getDatabaseFactory().setExecutor(new ForkJoinPool(30,
+                pool -> {
+                    final ForkJoinWorkerThread thread = ForkJoinPool.defaultForkJoinWorkerThreadFactory.newThread(pool);
+                    thread.setName("ConcurrentUpdatePool-" + threadCounter.getAndIncrement());
+                    return thread;
+                },
+                null, false));
+        final long seed = 320947L;
+        final boolean isGrouped = true;
+        // updating the parent & child concurrently is not thread safe, we may want to fix the behavior, or say that is
+        // not supported, as it is the same as updating the same record concurrently, which I don't think is generally
+        // supported.
+        final boolean isSynthetic = false;
+        final boolean primaryKeySegmentIndexEnabled = true;
+        // LucenePartitioner is not thread safe, and the counts get broken
+        // See: https://github.com/FoundationDB/fdb-record-layer/issues/2990
+        final int partitionHighWatermark = -1;
+        final LuceneIndexTestDataModel dataModel = new LuceneIndexTestDataModel.Builder(seed, this::getStoreBuilder, pathManager)
+                .setIsGrouped(isGrouped)
+                .setIsSynthetic(isSynthetic)
+                .setPrimaryKeySegmentIndexEnabled(primaryKeySegmentIndexEnabled)
+                .setPartitionHighWatermark(partitionHighWatermark)
+                .build();
+
+        final int repartitionCount = 10;
+        final int loopCount = 30;
+
+        final RecordLayerPropertyStorage contextProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_REPARTITION_DOCUMENT_COUNT, repartitionCount)
+                .addProp(LuceneRecordContextProperties.LUCENE_MAX_DOCUMENTS_TO_MOVE_DURING_REPARTITIONING, dataModel.nextInt(1000) + repartitionCount)
+                .addProp(LuceneRecordContextProperties.LUCENE_MERGE_SEGMENTS_PER_TIER, (double)dataModel.nextInt(10) + 2) // it must be at least 2.0
+                .build();
+        for (int i = 0; i < loopCount; i++) {
+            LOGGER.info(KeyValueLogMessage.of("concurrentUpdate loop",
+                    "iteration", i,
+                    "groupCount", dataModel.groupingKeyToPrimaryKeyToPartitionKey.size(),
+                    "docCount", dataModel.groupingKeyToPrimaryKeyToPartitionKey.values().stream().mapToInt(Map::size).sum(),
+                    "docMinPerGroup", dataModel.groupingKeyToPrimaryKeyToPartitionKey.values().stream().mapToInt(Map::size).min(),
+                    "docMaxPerGroup", dataModel.groupingKeyToPrimaryKeyToPartitionKey.values().stream().mapToInt(Map::size).max()));
+
+            long start = 234098;
```

Comment:
From a previous PR - shouldn't we change the start between iterations?

Reply:
I assume you're referring to this: https://github.com/FoundationDB/fdb-record-layer/pull/2992#discussion_r1878883222
I think I reasoned to myself there, that `start` was added to make the "timestamp" look more like a timestamp, and not just 0, 1, 2, ...
We auto increment based on the count in the group (plus some randomness), so this really could be 0.
I think it would be best to push `start` into a field in the data model, but perhaps as a later PR (I have some additional tests I've started adding on top of this)